### PR TITLE
Fix device normalization of automatically generate methods for custom backends. 

### PR DESCRIPTION
**Pull request by fakeYan**
Fixes #ISSUE_NUMBER
Fix the problem of error handling when the device input parameter adopts string type, Align capabilities.

`foo_storage = torch.DoubleStorage(4).foo(device="foo:0, non_blocking=False")`

cc @bdhirsh

### CODE REVIEW

File: torch/utils/backend_registration.py [hunk 1]
```diff
@@ -104,6 +104,12 @@ def _get_current_device_index():
             device_idx = _get_current_device_index()
         else:
             device_idx = device.index
+    # if isinstance(device, str), this means that the parameter passed in is in the string format "foo:0"
+    elif isinstance(device, str):
+        if device.split(":")[0] == custom_backend_name:
+            device_idx = int(device.split(":")[-1])
+        else:
+            raise RuntimeError(f"Invalid device, must be {custom_backend_name} device")
```
Comment (Skylion007) [inline torch/utils/backend_registration.py:112]:
```suggestion
            raise ValueError(f"Invalid device, must be {custom_backend_name} device")
```

**Review by Skylion007 — APPROVED** [review]

File: torch/utils/backend_registration.py [hunk 2]
```diff
@@ -104,6 +104,12 @@ def _get_current_device_index():
             device_idx = _get_current_device_index()
         else:
             device_idx = device.index
+    # if isinstance(device, str), this means that the parameter passed in is in the string format "foo:0"
+    elif isinstance(device, str):
+        if device.split(":")[0] == custom_backend_name:
```
Comment (Skylion007) [inline torch/utils/backend_registration.py:109]: Actually, hold on, what about `device="cuda"` or device="foo". Do you need to add error handing for that case as well?

**Review by Skylion007 — CHANGES_REQUESTED** [review]
Actually, hold on there might be a better way to do this. Why not just `torch.device(device)` to the conversion from str to device and then reuse the existing logic?

**Comment by fakeYan** [comment]
> Actually, hold on there might be a better way to do this. Why not just `torch.device(device)` to the conversion from str to device and then reuse the existing logic?

I understand the relevant logic. I moved the logic of converting str type to torch.device type to the front, so that objects of type torch.device can be processed normally later.  I think it looks better.

**Review by bdhirsh — APPROVED** [review]
lgtm, thanks!

**Pull request closed without merging**